flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -199,7 +252,32 @@ private void flush() throws InterruptedException {
      */
     private void completeRequest(Collection<RequestEntryT> failedRequestEntries) {
         inFlightRequestsCount--;
-        failedRequestEntries.forEach(bufferedRequestEntries::addFirst);
+        failedRequestEntries.forEach(failedEntry -> addEntryToBuffer(failedEntry, true));
+    }
+
+    private void addEntryToBuffer(RequestEntryT entry, boolean insertAtHead) {
+        if (bufferedRequestEntries.isEmpty() && !existsActiveTimerCallback) {
+            registerCallback();
+        }
+        if (insertAtHead) {
+            bufferedRequestEntries.addFirst(entry);
+        } else {
+            bufferedRequestEntries.add(entry);
+        }
+
+        double requestEntrySizeMB = getSizeInMB(entry);
+
+        if (insertAtHead) {
+            bufferedRequestEntriesSizeMB.addFirst(requestEntrySizeMB);
+        } else {
+            bufferedRequestEntriesSizeMB.add(requestEntrySizeMB);
+        }
+
+        bufferedRequestEntriesTotalSizeMB += requestEntrySizeMB;
+    }
+
+    private double getSizeInMB(RequestEntryT requestEntry) {
+        return getSizeInBytes(requestEntry) / (double) BYTES_IN_MB;
```

Comment:
Yeah I think we should just deal with `bytes` then we do not need to do any scaling. Also it removes the ambiguity around MB and MiB.

Reply:
Agreed!